README.md
# Table of Contents

* Table of Contents
* What is it?
* Quick Start
* Miscellaneous
* Key-value files
    * Configuration file
    * Problem descriptions

# What is it?

Bo Krif Ulse are the three Words of Power for the great shout of speedrunning; they translate roughly to "move, strive, forever". They reflect the goals of this tool: to practice precise, fast pill movements in the endless struggle to improve at Dr. Mario.

I have found that there are a couple of pill movements that come up repeatedly while playing Dr. Mario. I wanted a way to experiment with these movements, with various goals:

1. Sometimes there are multiple sensible paths to getting a pill to a particular location. Which one is fastest? It is tough to experiment with this during a game, because you must choose one and go, so you cannot easily try different variations, and because there is little in-game feedback about how long a particular pill drop actually took, so you cannot judge the relative merit of different variations.
2. In other situations, there is just one sensible path, but it is precise. But each individual maneuver might only come up infrequently, making it difficult to practice that maneuver. It would be nice to be able to pick a single maneuver and practice it repeatedly, until it becomes smooth, automatic, and consistent.
3. I wanted to learn the timing of various extremely common moves -- such as moving to a particular column before dropping a pill -- with some control over how long I practice a particular timing before moving on to another.
4. In some situations, I know there are two ways to do a thing, but I don't know which one is riskier. I'd like to be able to try it many times and see some statistics about how often I made a mistake, so that I can make informed decisions during gameplay about which moves to attempt and which to avoid.

This script helps with these goals by letting you set up a Dr. Mario board however you like, and repeatedly giving you a particular pill to place on that board. Once it's placed, the board is reset to its original state and you may immediately practice it again. It also reports some statistics; for now, these are how many frames you took to place the pill, whether you hit the right location or not, your success rate, and the average frame count for successful placements. Here's a short teaser showing these features:

![Teaser showing bokrifulse features](teaser/teaser.gif)

In the top left is the average successful frame count. Underneath the average, the frame counts for individual successes are listed. To the right is the fraction of attempts at which the pill locked in the right location, and below that (when appropriate) is the number of frames the last unsuccessful attempt took.

# Quick Start

1. Unzip [the latest release](http://dmwit.com/bokrifulse/00-bokrifulse-latest.zip) somewhere; I'll call the directory you chose DIR in the rest of these instructions.
2. Start [fceux](http://www.fceux.com/web/home.html) and load the Dr. Mario ROM.
3. Go to File &rarr; Load Lua Script, navigate to DIR, and choose bokrifulse.lua.
4. Start a 1 player game. I recommend choosing level 0, but the script will deal with it no matter what settings you choose.

# Miscellaneous

You may advance to the next problem either automatically by succeeding 15 times or manually by holding up while the pill locks.

# Key-value files

Bo Krif Ulse stores all of its configuration and runtime data in key-value files. These files are line-oriented; each line is either a comment (these start with `#`) or has a key and a value separated by an `=`. For example,

    # this is a comment

    dragons = cool

is a valid key-value file that associates the key `dragons` with the list of values [`cool`]; both the first line and the blank line following it count as comments and are ignored. Some values have to be in a specific format. A precise description of the syntax and special value formats follows.

Files are read in the latin-1 encoding because this is a lua script and I am lazy. A *whitespace character* is a tab, newline, vertical line feed, form feed, carriage return, or space (bytes 9, 10, 11, 12, 13, and 32). *Whitespace* is any sequence of whitespace characters (including the empty sequence). A *line break* is a newline, vertical line feed, form feed, or carriage return.

A *key* is any sequence of characters other than `=` or line breaks which does not start with `#` or a whitespace character and does not end with a whitespace character. A *value* is any sequence of characters other than line breaks which does not start or end with a whitespace character.

In this documentation, we use square brackets to enclose lists, and commas to separate values in the list. We use `monospace` for literal values and **bold** for metavariables standing in for a value; for example, [`foo`, **bar**] describes a list with two values in it, whose first value is exactly the string containing the characters `f`, `o`, and `o`, and whose second value can be referred to in surrounding text by the name **bar** but whose value we have not yet specified precisely.

The denotation of a key-value file is a total mapping from keys to lists of values. Each line of the file is expected to be in one of three formats:

1. Whitespace-only. These lines are ignored.
2. A comment, which is whitespace, followed by the character `#`, followed by any characters at all. These lines are ignored.
3. A key-value pair, which is whitespace, a key, whitespace, the character `=`, whitespace, a value, and whitespace. This adds the value to the list of values associated with the key.

Keys not mentioned in the file text are associated with [ ]. Multiple lines with the same key are allowed; the list of values associated with that key will be in the same order as they are in the file.

Below we will describe some *formats*, which assign special meanings to specific keys. For a particular format, a key may have a *default*, which is a value. If a key has default **d** in some format, then any key-value file whose denotation maps that key to [ ] will behave instead as if it maps that key to [**d**].

A *number* is a value in any format that lua's `tonumber` turns into a whole number. When giving bounds on these numbers, the bounds are inclusive. A *single number* is [**n**] where **n** is a number.

A *color character* is one of the characters `b`, `r`, `y`, `B`, `R`, or `Y`, with the following meaning:

| Character | Meaning |
|:---------:|:------- |
|`b`|blue|
|`B`|blue|
|`r`|red|
|`R`|red|
|`y`|yellow|
|`Y`|yellow|

## Configuration file

The top-level configuration must be stored in a file named `config` in the same directory as the `bokrifulse.lua` script itself. It is a key-value file, and gives the following keys special meaning:

* `version`: Must be [`1`].
* `problem`: The values in this list are treated as filenames. They should be relative paths, and will be interpreted as relative to the directory containing `bokrifulse.lua`. Each file referenced will be parsed as a problem file (see the subsection "Problem descriptions"); any invalid files will be skipped. If every file gets skipped, a special hard-coded problem will be used instead. The resulting non-empty list of problems will be cycled through one at a time; when the player succeeds 15 times, the script will advance to the next one.

## Problem descriptions

Individual problems describe the board position for the script to create, which pill should be presented to the user, and what target locations constitute a success. They are described in key-value files, and give the following keys special meaning:

* `version`: Must be [`1`].
* `row`: Must be a list containing exactly 16 values. The interpretation of these values is described below.
* `pill`: This specifies which pill Dr. Mario will throw into the bottle for the player. It must be [**pill**] and **pill** must be two color characters. The first color character describes the left half of the pill, and the second color character the right half.
* `goal x`: A single number between 0 and 7. The player must lock the pill with its bottom-left corner at this x position to be considered successful. 0 is the left-most column; 7 the right-most.
* `goal y`: A single number between 0 and 15. The player must lock the pill with its bottom-left corner at this y position to be considered successful. 0 is the bottom row; 15 the top.
* `goal orientations`: A non-empty list of numbers between 0 and 3. The player must lock the pill in one of the orientations in this list to be considered successful. The number describes how many counterclockwise rotations are needed to get from the initial pill that Dr. Mario threw to the desired orientation.
* `speed`: A single number between 0 and 49; default 0. The game will behave as if enough pills have dropped since the start of the level to increase the speed this many times.
* `viruses`: A single number between 1 and 99; default 99. This one is just for fun; it sets the virus count in the lower right of the screen to the given number.

The values for the `row` key are pictorial representations of the contents of the board that the player will be presented with. They use `--` for empty space, and a pair of characters describing the color and shape of a cell for the occupied spaces. For example, the string

    --rxrxbo--y<r>----

describes a row that has an empty space, two red viruses, a blue pill half, an empty space, a yellow and red pill, and two more empty spaces. The precise format description follows.

A *shape character* is one of the characters `x`, `o`, `<`, `>`, `V`, `v`, or `^`, with the following meaning:

| Character | Meaning |
|:---------:|:------- |
|`x`|virus|
|`o`|disconnected pill half|
|`<`|left half of a horizontal pill|
|`>`|right half of a horizontal pill|
|`V`|bottom half of a vertical pill|
|`v`|bottom half of a vertical pill|
|`^`|top half of a vertical pill|

A *cell* is two characters in one of the following forms:

* The exact string `--`. This indicates an unoccupied space on the playfield.
* A color character followed by a shape character. This indicates an occupied space of the given color and shape.
* Two hex digits (`0` through `9`, `a` through `f`, or `A` through `F`). The resulting byte will be used to index into Dr. Mario's sprite table.

A *playfield row* is a value containing exactly 8 cells separated by whitespace. The first cell describes the left-most column, the last the right-most column.

A *playfield* is a list of 16 playfield rows; the first playfield row describes the top of the screen, the last playfield row describes the bottom.

The `row` key must be associated with a playfield.


minefield/ChangeLog.md
# Revision history for minefield

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


minefield/Main.hs
{-# LANGUAGE LambdaCase #-}
module Main where

import Control.Applicative
import Control.Monad
import Control.Monad.IO.Class
import Data.Foldable
import Data.IORef
import Data.Set (Set)
import Dr.Mario.Model
import ListT (ListT(ListT))
import System.Random.MWC
import qualified ListT as L
import qualified Data.Set as S

data ButtonPress
	= Move Direction
	| Rotate Rotation
	deriving (Eq, Ord, Show)

-- only works for short lists, but that's fine, we're only passing in a 5-element list
unsafeShuffle :: MonadIO m => GenIO -> [a] -> m [a]
unsafeShuffle gen xs = do
	let n = length xs
	order <- liftIO (uniformR (0, product [2..n]-1) gen)
	pure (go n order xs)
	where
	go 0 _ _ = []
	go n order xs = x : go (n-1) q (b++e) where
		(q,r) = order `quotRem` n
		(b, x:e) = splitAt r xs

shuffle :: GenIO -> [a] -> IO [a]
shuffle gen xs = go (length xs) xs where
	go :: Int -> [a] -> IO [a]
	-- 11! fits in 25 bits, and Haskell promises Int holds at least 29 bits, so
	-- unsafeShuffle is safe
	go n xs | n <= 12 = unsafeShuffle gen xs
	go n xs = let n' = n-1 in do
		ix <- uniformR (0, n') gen
		let (b, x:e) = splitAt ix xs
		(x:) <$> go n' (b++e)

buttonPress :: GenIO -> ListT IO ButtonPress
buttonPress gen = do
	moves <- unsafeShuffle gen [Move left, Move right, Move down, Rotate Clockwise, Rotate Counterclockwise]
	asum (map pure moves)

onFailure :: Monad m => m () -> ListT m a -> ListT m a
onFailure failAct (ListT act) = ListT $ do
	v <- act
	case v of
		Nothing -> failAct
		_ -> pure ()
	pure v

paths :: GenIO -> IORef (Set Pill) -> Board -> Pill -> Pill -> ListT IO [Pill]
paths gen deadRef b tgt = go S.empty False where
	go visited die src
		| src == tgt = pure [src]
		| y (bottomLeftPosition src) < y (bottomLeftPosition tgt) = empty
		| src `S.member` visited = empty
		| otherwise = do
			dead <- liftIO (readIORef deadRef)
			guard (src `S.notMember` dead)
			onFailure (when die (modifyIORef deadRef (S.insert src))) $ do
				press <- buttonPress gen
				src' <- maybe empty pure $ case press of
					Move d -> move b src d
					Rotate r -> rotate b src r
				(src:) <$> go (S.insert src visited) (press == Move down) src'

chooseTarget :: GenIO -> IO Pill
chooseTarget gen = do
	v <- uniformR (0, 59) gen
	let (px, py) = (v `rem` 32) `quotRem` 4
	pure Pill
		{ content = PillContent
			{ orientation = if v < 32 then Vertical else Horizontal
			, bottomLeftColor = Yellow
			, otherColor = Yellow
			}
		, bottomLeftPosition = Position { x = px, y = py }
		}

minfectRB :: IOBoard -> Position -> IO ()
minfectRB mb p = minfect mb p (if parity (x p) == parity (y p) then Red else Blue) where
	parity x = (x `quot` 2) `rem` 2

positions :: [Pill] -> Set Position
positions ps = S.fromList
	[ pos
	| p <- ps
	, pos <- [bottomLeftPosition p, otherPosition p]
	, y pos < 16
	]

findAmbiguities :: [Pill] -> ListT IO [Pill] -> IO ([Pill], Set Position, Set Position)
findAmbiguities ps = go where
	poss = positions ps
	go pss = L.uncons pss >>= \case
		Nothing -> pure (ps, S.empty, S.empty)
		Just (ps', pss') -> if S.null unique && S.null unique'
			then go pss'
			else pure (ps', unique, unique')
			where
			poss' = positions ps'
			unique = poss S.\\ poss'
			unique' = poss' S.\\ poss

canonicalSrc = Pill
	{ content = PillContent
		{ orientation = Horizontal
		, bottomLeftColor = Yellow
		, otherColor = Yellow
		}
	, bottomLeftPosition = Position { x = 3, y = 15 }
	}

constrain :: GenIO -> IOBoard -> Pill -> IO ()
constrain gen mb tgt = do
	deadRef <- newIORef S.empty
	b <- mfreeze mb
	Just solution <- L.head (paths gen deadRef b tgt canonicalSrc)
	go deadRef solution
	where
	go deadRef solution = do
		b <- mfreeze mb
		(solution', unique, unique') <- findAmbiguities solution (paths gen deadRef b tgt canonicalSrc)
		case (S.size unique, S.size unique') of
			(0, 0) -> pure ()
			(n, 0) -> minfectRandom unique  >> go deadRef solution'
			_      -> minfectRandom unique' >> go deadRef solution

	minfectRandom ps = do
		ix <- uniformR (0, S.size ps-1) gen
		minfectRB mb (S.elemAt ix ps)

deleteSuperfluousViruses :: GenIO -> IOBoard -> Pill -> IO ()
deleteSuperfluousViruses gen mb tgt = do
	b <- mfreeze mb
	viruses <- shuffle gen
		[ pos
		| x <- [0..7]
		, y <- [0..15]
		, let pos = Position x y
		, unsafeGet b pos /= Empty
		]
	for_ viruses $ \pos -> do
		deadRef <- newIORef S.empty
		mclear mb [pos]
		b <- mfreeze mb
		Just (ps, pss) <- L.uncons (paths gen deadRef b tgt canonicalSrc)
		(_, unique, unique') <- findAmbiguities ps pss
		unless (S.null unique && S.null unique') (minfectRB mb pos)

ppDead :: Set Pill -> String
ppDead = foldMap $ \p -> concat
	[ show . x . bottomLeftPosition $ p
	, if orientation (content p) == Horizontal then "-" else "|"
	, reverse . take 2 . reverse . ("0"++) . show . y . bottomLeftPosition $ p
	]

bokrifulseFormat :: Board -> Pill -> String
bokrifulseFormat b tgt = unlines $
	[ "version = 1"
	, "pill = ry"
	, "goal x = " ++ show (x (bottomLeftPosition tgt))
	, "goal y = " ++ show (y (bottomLeftPosition tgt))
	, "goal orientation = " ++ if orientation (content tgt) == Vertical then "1" else "0"
	] ++
	[ "row = " ++ [c | x <- [0..7], c <- fmtCell (unsafeGet b (Position x y))]
	| y <- [15,14..0]
	]
	where
	fmtCell Empty = "--"
	fmtCell (Occupied color shape) = [fmtColor color, fmtShape shape]
	fmtColor Red = 'r'
	fmtColor Blue = 'b'
	fmtColor Yellow = 'y'
	fmtShape Virus = 'x'
	fmtShape Disconnected = 'o'
	fmtShape North = '^'
	fmtShape South = 'v'
	fmtShape East = '<'
	fmtShape West = '>'

main :: IO ()
main = createSystemRandom >>= \gen -> do
	tgt <- chooseTarget gen
	mb <- memptyBoard 8 16
	constrain gen mb tgt
	deleteSuperfluousViruses gen mb tgt
	b <- munsafeFreeze mb
	putStr (bokrifulseFormat b tgt)


bokrifulse.lua
local function get_prefix()
	local dir_sep = package.config:sub(1,1)
	local script_dir = debug.getinfo(2, "S").source:match('.(.*' .. dir_sep .. ')')
	if script_dir == nil then
		return '.' .. dir_sep
	else
		return script_dir
	end
end

local prefix = get_prefix()
local config_filename = prefix .. 'config'
local default_problem =
	{ board =
		{ 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0x60, 0x70, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xd0, 0xd0, 0xff, 0xff, 0xff
		, 0xff, 0xff, 0xff, 0xd0, 0xd0, 0xff, 0xff, 0xff
		}
	, bcd_virus_count = 0x99
	, pill_left = 0
	, pill_right = 0
	, speed_adjustment = 0
	, goal_x = 3
	, goal_y = 3
	, goal_orientations = { 0, 2 }
	, source = "bokrifulse.lua"
	}
local PROBLEMS = { default_problem }
local CURRENT_PROBLEM = 1
local SUCCESSES = {}
local FAILURE_COUNT = 0
local LAST_FAILURE = nil
local LAST_FAILURE_FRAME = nil

-- Where you would write
--     x <- act
--     foo
-- in Haskell, write
--     bind(act, function(x)
--     foo
-- in Lua. Where you would write
--     unless p (throwError e)
-- in Haskell, write
--     bind({p, e}, function()
-- in Lua. Only wrinkle: however many times you bind in your "do block", you
-- need that many copies of end) afterwards.
local function bind(act, f)
	if act[1] then return f(act[2]) else return act end
end

local function forM(kvs, f)
	local RESULT = {}
	for k, v in pairs(kvs) do
		local fv = f(v)
		if fv[1] then RESULT[k] = fv[2] else return fv end
	end
	return {true, RESULT}
end

local function to_bcd(n)
	return n % 10 + 8 * (n - n % 10) / 5
end

local function from_bcd(n)
	return n % 16 + 5 * (n - n % 16) / 8
end

local function parse_file(filename)
	local HANDLE, err_description = io.open(filename, "r")
	if HANDLE == nil then
		return {false, ("Could not open file for reading (%s)"):format(err_description)}
	end

	local RESULT = {}
	local SUCCESS = true
	local LINE_NUMBER = 1

	while SUCCESS do
		local line = HANDLE:read("*line")
		if line == nil then break end

		if line:find("^%s*$") == nil and line:find("^%s*#.*$") == nil then
			k, v = line:match("^%s*(.-)%s*=%s*(.-)%s*$")

			if k == nil then
				SUCCESS = false
				RESULT = 'invalid syntax on line ' .. LINE_NUMBER
			elseif RESULT[k] == nil then
				RESULT[k] = { v }
			else
				table.insert(RESULT[k],v)
			end

			LINE_NUMBER = LINE_NUMBER+1
		end
	end

	HANDLE:close()
	return {SUCCESS, RESULT}
end

local function check_version(kvs, expected_version)
	local version_list = kvs['version'] or {}
	return bind({#version_list == 1, ('Expected exactly one version (saw %d)'):format(#version_list)}, function()
	return bind({version_list[1] == expected_version, ('Expected version %s (saw %s)'):format(expected_version, version_list[1])}, function()
	return {true, kvs}
	end) end)
end

local function parse_versioned_file(filename, version)
	return bind(parse_file(filename), function(kvs)
	return check_version(kvs, version)
	end)
end

local color_lower_nibble =
	{   y   = 0x00
	,   Y   = 0x00
	,   r   = 0x01
	,   R   = 0x01
	,   b   = 0x02
	,   B   = 0x02
	}
local shape_upper_nibble =
	{ ['^'] = 0x40
	,   v   = 0x50
	,   V   = 0x50
	, ['<'] = 0x60
	, ['>'] = 0x70
	,   o   = 0x80
	,   x   = 0xd0
	}

local function parse_cell(s)
	if s == '--' then
		return {true, 0xff}
	elseif s:find('[BbRrYy][xo><Vv^]') then
		local byte = color_lower_nibble[s:sub(1,1)] + shape_upper_nibble[s:sub(2,2)]
		return {true, byte}
	elseif s:find('%x%x') then
		return {true, tonumber('0x' .. s)}
	else
		return {false, ('Expected a cell, namely --, two hex digits, or a color in r, y, b and a shape in x, o, ^, v, <, > (saw %s)'):format(s)}
	end
end

local function parse_problem_row(row)
	local I = 1
	return forM({1,2,3,4,5,6,7,8}, function(j)
		I = row:find('%S%S',I)
		return bind({I ~= nil, ('Expected 8 cells (saw %d in %s)'):format(j-1, row)}, function()
		return bind({j < 8 or I+1 == row:len(), ('Expected 8 cells (saw more in %s)'):format(row)}, function()
		I = I+2
		return parse_cell(row:sub(I-2,I-1))
		end) end)
	end)
end

local function parse_number(s, k, lo, hi)
	local n = tonumber(s)
	return bind({n ~= nil and lo <= n and n <= hi and math.floor(n) == n, ('Expected an integer in the range %d-%d in %s (saw %s)'):format(lo, hi, k, s)}, function()
	return {true, n}
	end)
end

local function parse_single_number(kvs, k, def, lo, hi)
	local v = kvs[k] or def
	return bind({#v == 1, ('Expected exactly one %s (saw %d)'):format(k, #v)}, function()
	return parse_number(v[1], k, lo, hi)
	end)
end

local function parse_pill(vs)
	return bind({#vs == 1, ('Expected exactly one pill (saw %d)'):format(#vs)}, function()
	local v = vs[1]
	return bind({v:find('^[rRyYbB][rRyYbB]$'), ('Expected two colors drawn from r, y, b (saw %s)'):format(v)}, function()
	return {true, {color_lower_nibble[v:sub(1,1)], color_lower_nibble[v:sub(2,2)]}}
	end) end)
end

local function parse_orientations(vs)
	return bind({#vs >= 1, ('Expected at least one goal orientation (saw %d)'):format(#vs)}, function()
	return forM(vs, function(v) return parse_number(v, 'goal orientation', 0, 3) end)
	end)
end

local function concat(xss)
	local YS = {}
	local I = 1
	for _, xs in pairs(xss) do
		for _, x in pairs(xs) do
			YS[I] = x
			I = I + 1
		end
	end
	return YS
end

local function parse_problem(filename)
	return bind(parse_versioned_file(filename, '1'), function(kvs)
	local row_count = #(kvs.row or {})
	return bind({row_count, ('Expected exactly 16 rows (saw %d)'):format(row_count)}, function()
	return bind(forM(kvs.row, parse_problem_row), function(rows)
	return bind(parse_single_number(kvs, 'viruses', {'99'}, 1, 99), function(virus_count)
	return bind(parse_pill(kvs.pill or {'ry'}), function(pill)
	return bind(parse_single_number(kvs, 'speed', {'0'}, 0, 49), function(speed_adjustment)
	return bind(parse_single_number(kvs, 'goal x', {}, 0, 7), function(goal_x)
	return bind(parse_single_number(kvs, 'goal y', {}, 0, 15), function(goal_y)
	return bind(parse_orientations(kvs['goal orientation'] or {}), function(goal_orientations)
	return {true,
		{ board = concat(rows)
		, bcd_virus_count = to_bcd(virus_count)
		, pill_left = pill[1]
		, pill_right = pill[2]
		, speed_adjustment = speed_adjustment
		, goal_x = goal_x
		, goal_y = goal_y
		, goal_orientations = goal_orientations
		}}
	end) end) end) end) end) end) end) end) end)
end

local function parse_config(filename)
	local PROBLEMS = {}
	return bind(parse_versioned_file(filename, '1'), function(kvs)

	for _, problem_suffix in pairs(kvs.problem or {}) do
		local problem_filename = prefix .. problem_suffix
		local PROBLEM = parse_problem(problem_filename)
		if PROBLEM[1] then
			PROBLEM[2].source = problem_suffix
			table.insert(PROBLEMS, PROBLEM[2])
		else
			print('Skipping invalid problem description in ' .. problem_filename .. ':\n' .. PROBLEM[2])
		end
	end

	return {true, { problems = PROBLEMS }}
	end)
end

local function reload_config()
	local config = parse_config(config_filename)

	PROBLEMS = {}
	if config[1] then
		PROBLEMS = config[2].problems
	else
		print('Skipping invalid config file ' .. config_filename .. ':\n' .. config[2])
	end

	if #PROBLEMS == 0 then
		print('No valid problems specified in ' .. config_filename .. '.\nUsing a hardcoded default problem as a fallback.')
		table.insert(PROBLEMS, default_problem)
	end
end

local function change_problem()
	reload_config()
	SUCCESSES = {}
	FAILURE_COUNT = 0
	CURRENT_PROBLEM = CURRENT_PROBLEM + 1
	if CURRENT_PROBLEM > #PROBLEMS then
		CURRENT_PROBLEM = 1
	end
end

local success_color = "#287c0e"
local failure_color = "#790c0c"
local bg_color = "#b2bdff"
local num_width_px = 6
local num_height_px = 9
local gui_padding_x_px = 9
local gui_padding_y_px = 9
local gui_max_x = 256
local gui_max_y = 224
local failure_display_time = 90 -- in frames

local function render()
	local success_count = #SUCCESSES

	if success_count > 0 then
		local SUM = 0
		for i, n in pairs(SUCCESSES) do
			SUM = SUM + n
			gui.text(gui_padding_x_px, gui_padding_y_px + num_height_px*(i+1), ("%3d"):format(n), success_color, bg_color)
		end
		gui.text(gui_padding_x_px, gui_padding_y_px + num_height_px, ("%6.2f"):format(SUM/success_count), success_color, bg_color)
	end

	if success_count + FAILURE_COUNT > 0 then
		local success_ratio = success_count / (success_count + FAILURE_COUNT)
		gui.text(2*gui_padding_x_px + 6*num_width_px, gui_padding_y_px + num_height_px, ("%3.0f%%"):format(100*success_ratio), success_color, bg_color)
	end

	if LAST_FAILURE_FRAME then
		gui.text(2*gui_padding_x_px + 6*num_width_px, gui_padding_y_px + 2*num_height_px, ("%3d"):format(LAST_FAILURE), failure_color, bg_color)
		if LAST_FAILURE_FRAME + failure_display_time < emu.framecount() then
			LAST_FAILURE = nil
			LAST_FAILURE_FRAME = nil
		end
	end

	gui.text(gui_padding_x_px, gui_max_y - gui_padding_y_px, PROBLEMS[CURRENT_PROBLEM].source)
end

local initial_state = 0 -- waiting for the user to start the game
local virus_generation_state = 1 -- the game is picking random viruses, let's politely wait for it to finish
local ready_state = 2 -- our turn: time modify the board and pill lookahead however we want
local awaiting_control_state = 3 -- we have set up the board and pill lookahead appropriately, now we're waiting for the pill animation to complete and the pill to appear at the top of the board
local under_control_state = 4 -- the player is doing maneuvers
local maybe_maybe_locked_state = 5 -- the pill is about to drop one row
local maybe_locked_state = 6 -- the pill just dropped one row, and may be either locked about to drop another row
local locked_state = 7 -- the pill locked two frames ago
local CURRENT_STATE = initial_state
local MANEUVER_START_FRAME

local function initial_action()
	local mode = memory.readbyte(0x46)
	local player_count = memory.readbyte(0x727)
	local level = math.min(memory.readbyte(0x316), 20)
	local bcd_virus_count = memory.readbyte(0x324)
	local max_virus_count = (level+1)*4
	-- virus_count < max_virus_count is important: we want to wait until
	-- they're equal, but sometimes they are spuriously equal when the number
	-- of viruses is left over unchanged from a previous game
	if mode == 8 and player_count == 1 and bcd_virus_count < to_bcd(max_virus_count) then
		CURRENT_STATE = virus_generation_state
	end
end

local function virus_generation_action()
	local level = math.min(memory.readbyte(0x316), 20)
	local bcd_virus_count = memory.readbyte(0x324)
	local max_virus_count = (level+1)*4
	if bcd_virus_count == to_bcd(max_virus_count) then
		CURRENT_STATE = ready_state
	end
end

local function ready_action()
	local problem = PROBLEMS[CURRENT_PROBLEM]
	for i, v in pairs(problem.board) do
		memory.writebyte(0x400 + i - 1, v)
	end
	memory.writebyte(0x324, problem.bcd_virus_count)
	for i = 0, 127 do
		memory.writebyte(0x780 + i, 3*problem.pill_left + problem.pill_right)
	end
	memory.writebyte(0x31a, problem.pill_left)
	memory.writebyte(0x31b, problem.pill_right)
	memory.writebyte(0x30a, problem.speed_adjustment)
	CURRENT_STATE = awaiting_control_state
end

local function awaiting_control_action()
	local drop = memory.readbyte(0x312)
	if drop > 0 then
		MANEUVER_START_FRAME = emu.framecount()
		CURRENT_STATE = under_control_state
	end
end

local function under_control_action()
	local drop = memory.readbyte(0x312)
	if drop == 0 then
		CURRENT_STATE = maybe_maybe_locked_state
	end
end

local function maybe_maybe_locked_action()
	local drop = memory.readbyte(0x312)
	if drop == 0 then
		CURRENT_STATE = maybe_locked_state
	else
		CURRENT_STATE = under_control_state
	end
end

local function maybe_locked_action()
	local drop = memory.readbyte(0x312)
	if drop == 0 then
		CURRENT_STATE = locked_state
	else
		CURRENT_STATE = under_control_state
	end
end

local function locked_action()
	local problem = PROBLEMS[CURRENT_PROBLEM]
	local pill_x = memory.readbyte(0x305)
	local pill_y = memory.readbyte(0x306)
	local pill_orientation = memory.readbyte(0x325)
	-- This is not quite right; it could be off by one frame if a pill
	-- naturally drops one row, then gets forced to drop a row by the player on
	-- the very next frame, and the forced drop locks it. Currently, since we
	-- set the drop counter bonus index to 0 (and so natural drops always
	-- happen after an even number of frames), this can only occur if all of
	-- the following happen:
	--
	-- * The pill comes under player control on an odd frame.
	-- * The player lets it drop naturally until the second-to-last row before
	--   it locks.
	-- * On the exact frame that the pill transitions from the second-to-last
	--   row to the last row, the player presses down.
	--
	-- This combination seems very uncommon to me, especially given that I
	-- expect the way most people will use this tool is to hold down as much as
	-- they can. So we'll just report a one-frame-incorrect answer in the
	-- vanishingly unlikely case that all the conditions above are met.
	local maneuver_frames = emu.framecount() - 3 - MANEUVER_START_FRAME + 1

	local SUCCESS = false
	for _, goal_orientation in pairs(problem.goal_orientations) do
		SUCCESS = SUCCESS or pill_orientation == goal_orientation
	end
	SUCCESS = SUCCESS and pill_x == problem.goal_x and pill_y == problem.goal_y

	local CHANGE_PROBLEM = joypad.get(1)["up"]
	if SUCCESS then
		local success_count = #SUCCESSES + 1
		SUCCESSES[success_count] = maneuver_frames
		CHANGE_PROBLEM = CHANGE_PROBLEM or success_count >= 15
	else
		FAILURE_COUNT = FAILURE_COUNT + 1
		LAST_FAILURE = maneuver_frames
		LAST_FAILURE_FRAME = emu.framecount()
	end
	if CHANGE_PROBLEM then change_problem() end

	CURRENT_STATE = ready_state
end

local actions =
	{ [initial_state] = initial_action
	, [virus_generation_state] = virus_generation_action
	, [ready_state] = ready_action
	, [awaiting_control_state] = awaiting_control_action
	, [under_control_state] = under_control_action
	, [maybe_maybe_locked_state] = maybe_maybe_locked_action
	, [maybe_locked_state] = maybe_locked_action
	, [locked_state] = locked_action
	}

reload_config()
emu.registerbefore(function()
	actions[CURRENT_STATE]()
	render()
end)
